### PR TITLE
chore(RHTAPWATCH-634): request ODCS compose

### PR DESCRIPTION
**Pull request by Omeramsc**
Implement ODCSRequestor, allowing requesting ODCS
for a compose file.

### CODE REVIEW

**Comment by yftacherzog** [comment]
@Omeramsc changed to draft so that others won't spend time reviewing it until we made the changes we discussed this morning.

File: tests/test_odcs_requester.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+            id="single compose, tag source, no additional arguments",
+        ),
+        pytest.param(
+            ODCSComposesConfigs(
+                [
+                    ODCSComposeConfig(spec=ComposeSourceTag(tag="tag")),
+                    ODCSComposeConfig(spec=ComposeSourceTag(tag="tag")),
+                ]
+            ),
+            2,
```
Comment (gbenhaim) [inline tests/test_odcs_requester.py:60]: This can be `len(composes_configs.configs)` added to line 74

File: generate_compose/odcs_requester.py [hunk 2]
```diff
@@ -16,9 +19,21 @@ class ODCSRequestReference(ComposeReference):
...
+    Request a new ODCS composes based on compose configurations and return a reference
     to the remote compose location.
     """
 
-    def __call__(self, configs: ComposeConfigurations) -> ODCSRequestReference:
-        raise NotImplementedError()
+    odcs: ODCS = ODCS("https://odcs.engineering.redhat.com/")
+
+    def __call__(self, compose_configs: ComposeConfigurations) -> ODCSRequestReferences:
+        assert isinstance(compose_configs, ODCSComposesConfigs)
```
Comment (gbenhaim) [inline generate_compose/odcs_requester.py:29]: why do we need the type assertion? it break the abstraction of the protocol

File: generate_compose/odcs_requester.py [hunk 3]
```diff
@@ -16,9 +19,21 @@ class ODCSRequestReference(ComposeReference):
...
+        composes = [
+            self.odcs.request_compose(config.spec, **config.additional_args)
```
Comment (gbenhaim) [inline generate_compose/odcs_requester.py:31]: we would need a followup pr to add some calls to a logger to show what's is going on

File: tests/test_odcs_requester.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    num_of_composes: int,
+) -> None:
+    """test ODCSRequester.__call__"""
+    expected_req_ref = ODCSRequestReferences(
+        compose_urls=[compose_url] * num_of_composes
+    )
+    mock_odcs = create_odcs_mock(num_of_composes)
+    odcs_requester = ODCSRequester(odcs=mock_odcs)
+    req_ref = odcs_requester(compose_configs=composes_configs)
+
```
Comment (gbenhaim) [inline tests/test_odcs_requester.py:79]: we should assert that `wait_for_compose` was called as expected

File: tests/test_odcs_requester.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    assert mock_odcs.request_compose.call_count == num_of_composes
+    assert mock_odcs.request_compose.call_args[0][0].source == compose_source_tag.source
```
Comment (gbenhaim) [inline tests/test_odcs_requester.py:81]: I think that this should be replaced with `call_arg_list` that captures all the calls to the mock object.
In addition, I think that for generating the expected calls, you should use `composes_configs`.
`compose_source_tag` can be removed from this test, I don't see why we need it and `compose_configs` at the same time.

File: tests/test_odcs_requester.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    """Example compose-url, as close to the one expected"""
+    return "http://download.eng.bos.redhat.com/odcs/prod/odcs-222222"
+
+
+@pytest.fixture()
+def create_odcs_mock(compose_url: str) -> Callable[[int], MagicMock]:
+    """Create an ODCS mock with specific results for the compose method"""
+
+    def _mock_odcs_compose(num_of_composes: int) -> MagicMock:
+        """Monkey-patched ODCS compose"""
```
Comment (gbenhaim) [inline tests/test_odcs_requester.py:28]: I don't see any monky patch in here (which is good), so the doc strings should be updated.

File: tests/test_odcs_requester.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+                    ODCSComposeConfig(spec=ComposeSourceTag(tag="tag")),
+                    ODCSComposeConfig(spec=ComposeSourceTag(tag="tag")),
+                ]
+            ),
+            2,
+            id="multiple composes, , tag source, no additional arguments",
+        ),
+    ],
+)
+def test_odcs_requester(
```
Comment (gbenhaim) [inline tests/test_odcs_requester.py:65]: We need an additional test case for the negative path. Ensure that an exception is raised when the request for a compose fails or timing out when waiting for it.

**Review by gbenhaim — CHANGES_REQUESTED** [review]
Great progress! I've added some comments.

**Comment by ifireball** [comment]
@Omeramsc you need to rebase this to get a build from RHTAP

File: tests/test_odcs_requester.py (hunk 1 above)
Comment (Omeramsc) [reply tests/test_odcs_requester.py:60]: That's what I started with, I don't mind using it. I thought the parametrize will be more explicit.

File: tests/test_odcs_requester.py (hunk 4 above)
Comment (Omeramsc) [reply tests/test_odcs_requester.py:79]: We already assert it is being called. do we want to assert it was called with the specific ID? (mocked as the index of the compose)

File: tests/test_odcs_requester.py (hunk 6 above)
Comment (Omeramsc) [reply tests/test_odcs_requester.py:28]: Will fix, but personally I think we need to disable pylint's demand to include docstrings in`test_*` files. they should be self-explanatory.

Comment (ifireball) [reply tests/test_odcs_requester.py:28]: there is a pytest plugin for that - you can add it in a separate PR and then rebase this...

File: generate_compose/odcs_requester.py (hunk 2 above)
Comment (yftacherzog) [reply generate_compose/odcs_requester.py:29]: @gbenhaim what's the way around it? If we don't do this assertion, then the static checks will later on complain about accessing fields which are not part of the parent class.

File: tests/test_odcs_requester.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
+"""test_odcs_requester.py - test odcs_requester"""
+from typing import Callable
+
+# pylint: disable=redefined-outer-name
```
Comment (yftacherzog) [inline tests/test_odcs_requester.py:4]: What is this one addressing?
If I recall correctly, if we have disable comments on a line of their own, they will apply to everything below that point in the code. We should instead have them inline (I think), which will only make them apply to the specific line in which they are introduced.

Comment (Omeramsc) [reply tests/test_odcs_requester.py:4]: It's intended for each time a fixture functions is being passed as a parameter
for instance:
```
def test_odcs_requester(
    create_odcs_mock: Callable,
    compose_url: str,
   ```
   
   I placed it at the top of the file as, same with docstring, I think we should cancel it for tests.

File: tests/test_odcs_requester.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+            {"result_repofile": f"{compose_url}", "id": i}
+            for i in range(1, num_of_composes + 1)
+        ]
+        return mock
+
+    return _mock_odcs_compose
+
+
+@pytest.fixture()
+def compose_source_tag() -> ComposeSourceTag:
```
Comment (yftacherzog) [inline tests/test_odcs_requester.py:40]: I think we should have a mock here of the generic type rather than an instance of one of its implementations.

File: tests/test_odcs_requester.py (hunk 8 above)
Comment (yftacherzog) [reply tests/test_odcs_requester.py:4]: Do we need it to be a fixure?

Comment (Omeramsc) [reply tests/test_odcs_requester.py:4]: I'm not sure I'm following. what do you want to be a function, and why not use a fixture.

Comment (yftacherzog) [reply tests/test_odcs_requester.py:4]: What do we gain by making this mock-generating function a fixture. The reason not to make it a fixture is that it's (needlessly?) redefines an outer name. Also, if you don't make it a fixture, you could just use it without having it as an input to the test function. I think the test code could stay and we just drop the fixture name from the test inputs.

Am I missing something?

File: tests/test_odcs_requester.py (hunk 5 above)
Comment (Omeramsc) [reply tests/test_odcs_requester.py:81]: Thing is, when we call `request_compose` we give it a compose source as `spec`. So if we are using `call_args_list`, we will have to compare between the same objects (in memory), while we don't have access to the one created in the parametrization.  
we can instead either have separated tests for the different test-cases, or generate the objects in the test function according to the number of composes needed for each test, but I think it is not as good of a long-term solution (if we'd like to add test cases for different compose sources, or with different kwargs)

File: tests/test_odcs_requester.py (hunk 9 above)
Comment (Omeramsc) [reply tests/test_odcs_requester.py:40]: Maybe a fixture factory? either way we'll need to supply it with the `source_type`

Comment (yftacherzog) [reply tests/test_odcs_requester.py:40]: If we create a mock using `create_autospec` based on the the generic source, then I don't think we need anything.

File: tests/test_odcs_requester.py (hunk 8 above)
Comment (Omeramsc) [reply tests/test_odcs_requester.py:4]: We don't really redefine anything, This is the usual way fixtures are to be used... this error will repeat every time we use a fixture in a test function, same error repeats for the other fixtures [1]. I'm pretty sure it was suppressed in `product-ci-cd`, which is why we are not used to see it. 
Why not having it as a fixture? we might want to use it in other tests (for instance, the negative one Gal mentioned).
You cannot drop the fixture name from the test inputs, fixtures need to be mentioned in the test's inputs and not called directly; otherwise you'l get this error [2], which is an issue. 

[1]
```
tests/test_odcs_requester.py:23:21: W0621: Redefining name 'compose_url' from outer scope (line 17) (redefined-outer-name)
tests/test_odcs_requester.py:72:4: W0621: Redefining name 'create_odcs_mock' from outer scope (line 23) (redefined-outer-name)
tests/test_odcs_requester.py:73:4: W0621: Redefining name 'compose_url' from outer scope (line 17) (redefined-outer-name)
tests/test_odcs_requester.py:74:4: W0621: Redefining name 'compose_source' from outer scope (line 48) (redefined-outer-name)
```

[2]
```
Failed: Fixture "create_odcs_mock" called directly. Fixtures are not meant to be called directly
```

Comment (yftacherzog) [reply tests/test_odcs_requester.py:4]: This is not the first fixture to be used on this repo, and we're not getting this error for any other fixture, so it's not an issue for just any fixture.

File: generate_compose/odcs_requester.py [hunk 10]
```diff
@@ -16,9 +19,21 @@ class ODCSRequestReference(ComposeReference):
 @dataclass(frozen=True)
 class ODCSRequester(ComposeRequester):
     """
-    Request a new ODCS compose based on compose configurations and return a reference
+    Request a new ODCS composes based on compose configurations and return a reference
```
Comment (yftacherzog) [inline generate_compose/odcs_requester.py:22]:
```suggestion
    Request new ODCS composes based on compose configurations and return a reference
```

File: tests/test_odcs_requester.py (hunk 7 above)
Comment (Omeramsc) [reply tests/test_odcs_requester.py]: Added two tests- one where request compose fails (HTTPError), and another where wait_for_compose fails(RuntimeError)

File: tests/test_odcs_requester.py (hunk 6 above)
Comment (Omeramsc) [reply tests/test_odcs_requester.py]: will do

File: tests/test_odcs_requester.py (hunk 5 above)
Comment (Omeramsc) [reply tests/test_odcs_requester.py]: replaced

File: tests/test_odcs_requester.py (hunk 4 above)
Comment (Omeramsc) [reply tests/test_odcs_requester.py]: added an assertion for the specific IDs

File: tests/test_odcs_requester.py (hunk 1 above)
Comment (Omeramsc) [reply tests/test_odcs_requester.py]: Changed

File: tests/test_odcs_requester.py (hunk 9 above)
Comment (Omeramsc) [reply tests/test_odcs_requester.py]: changed

File: generate_compose/odcs_requester.py (hunk 10 above)
Comment (Omeramsc) [reply generate_compose/odcs_requester.py]: fixed

File: tests/test_odcs_requester.py (hunk 8 above)
Comment (Omeramsc) [reply tests/test_odcs_requester.py]: Test class solved it for this one. not sure it'll act the same way when not using test classes.

**Comment by yftacherzog** [comment]
@Omeramsc can you verify that you're able to use the code you wrote here to request a compose of some sort? You can use the equivalent inputs to what we have here: https://github.com/redhat-appstudio/tools/blob/main/generate_compose/odcs_ping.py#L33-L34

File: generate_compose/odcs_requester.py (hunk 2 above)
Comment (gbenhaim) [reply generate_compose/odcs_requester.py:29]: define a method/property on the protocol that provides you what you need.

Comment (yftacherzog) [reply generate_compose/odcs_requester.py:29]: Can you elaborate? We want to be able to use a property of the sub class instance (`configs`) that is not a part of the protocol.

**Review by gbenhaim — APPROVED** [review]

**Comment by Omeramsc** [comment]
> @Omeramsc can you verify that you're able to use the code you wrote here to request a compose of some sort? You can use the equivalent inputs to what we have here: https://github.com/redhat-appstudio/tools/blob/main/generate_compose/odcs_ping.py#L33-L34 You'd need to kinit before

Verified working
http://download.eng.bos.redhat.com/odcs/prod/odcs-2628389/compose/Temporary/odcs-2628389.repo
